File: problems/003_longest_substring_without_repeating_characters/solution.cpp
// Leetcode 3. Longest Substring Without Repeating Characters
// @tag: hashtalble, string, sliding-window
// @difficulty: medium
//
// Solution
// Two pointers

#if defined(SLIDING_WINDOW)
#include <algorithm>
#include <string>
#include <vector>

class Solution {
 public:
  int lengthOfLongestSubstring(const std::string& s) {
    if (s.empty()) return 0;

    int max_len = 0;
    const int len = static_cast<int>(s.size());
    std::vector<int> freq(128, 0);

    int left = 0;
    for (int right = 0; right < len; ++right) {
      ++freq[s[right]];
      while (freq[s[right]] > 1) {
        --freq[s[left]];
        ++left;
      }
      max_len = std::max(max_len, right - left + 1);
    }

    return max_len;
  }
};

// Better solution
// Can jump left directly to the next posisiton
#if defined(DP)
#include <algorithm>  // for std::min
#include <vector>

class Solution {
 public:
  int lengthOfLongestSubstring(string s) {
    if (s.size() == 0 || s.size() == 1) {
      return s.size();
    }
    int left = 0;
    int max_len = 0;
    std::vector<int> last_seen(128, -1);

    for (int right = 0; right < static_cast<int>(s.size()); ++right) {
      const char c = s[right];
      if (last_seen[c] != -1 && last_seen[c] >= left) {
        left = last_seen[c] + 1;
      }
      last_seen[c] = right;
      max_len = std::max(max_len, right - left + 1);
    }
    return max_len;
  }
};
#endif

File: problems/003_longest_substring_without_repeating_characters/Solution.hs
-- Leetcode 3. Longest Substring Without Repeating Characters
--
-- Algorithm: Sliding window (linear iterative process, SICP §1.2.1)
-- Time: O(n), Space: O(charset)
--
-- SICP Perspective:
--   - LINEAR ITERATIVE PROCESS: State evolves through each character
--   - State: (maxLen, leftBound, char→lastSeenIndex)
--   - Invariant: All chars in [left, i] are unique

module Solution where

import qualified Data.Map.Strict as M
import Data.List (fold')

-- | Returns length of longest substring without repeating characters.
-- Maintains sliding window [left, i] with uniqueness invariant.
lengthOfLongestSubstring :: String -> Int
lengthOfLongestSubstring s = best
  where
    -- State: (maxLen, leftBound, char→lastIndex)
    (best, _, _) = foldl' step (0, 0, M.empty) (zip [0..] s)

    step :: (Int, Int, M.Map Char Int) -> (Int, Char) -> (Int, Int, M.Map Char Int)
    step (best, left, mp) (i, c) =
      let -- Move left past duplicate (max needed for stale entries in map)
          left' = case M.lookup c mp of
            Just j -> max left (j + 1)
            Nothing -> left
          mp' = M.insert c i mp
          best' = max best (i - left' + 1)
       in (best', left', mp')

{- Test cases:
lengthOfLongestSubstring ""              -- 0
lengthOfLongestSubstring "a"             -- 1
lengthOfLongestSubstring "bbbbb"         -- 1
lengthOfLongestSubstring "abcabcbb"      -- 3   -- "abc"
lengthOfLongestSubstring "pwwkew"        -- 3   -- "wke"
lengthOfLongestSubstring "dvdf"          -- 3   -- "vdf"
lengthOfLongestSubstring "abba"          -- 2   -- "ab" or "ba" (why max needed)
-}
